### PR TITLE
Table returns TableRow instead of list of strings

### PR DESCRIPTION
**Pull request by olgamelnichuk**

### CODE REVIEW

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt [hunk 1]
```diff
@@ -7,26 +7,45 @@ import ac.uk.ebi.biostd.submission.File
...
+        return name == that!!.name
     }
+
+    override fun hashCode(): Int {
+        return name.hashCode()
+    }
+}
+
+data class TableRow(val id: String, val idPropertyName: String, val values: List<Attribute?>) {
+    fun valueList(): List<String> = listOf(id) +
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: I do not see `idPropertyName` used here or in any place we can safely removed it.

Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: I'd like to.. however in the json serialization we still need to know how id field is called. E.g.
links: [[ {url: '12345', attributes: [...]} ]]

Or we need to find an other way to do this..

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt [hunk 2]
```diff
@@ -7,26 +7,45 @@ import ac.uk.ebi.biostd.submission.File
...
-    fun getHeaders(): List<String> {
-        return headers.toList()
+    fun getHeaders(): List<TableHeader> {
+        return listOf(TableHeader(idHeaderName)) + headers.toList()
     }
 
     fun getRows(): List<TableRow> {
-        return rows.map { headers.map { header -> it.getOrDefault(header, NO_VALUE) } }
+        return elements.map { element ->
+            TableRow(element.getId(), element.getIdPropertyName(),
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: I do not like to lets scape null references  so here basically with retrieve a list with null elements and then latter in `TableRow` we handle this. I said lets solve problem from root adding a empy attribute `companion object`. Also private function can hel make code more understandable IMHO.

```
fun getRows(): List<TableRow> {
        return elements.map {
            TableRow(it.getId(), it.getIdPropertyName(), headers.map { header -> searchValue(header.name, it.getAttributes()) })
        }
    }
```

```
 private fun searchValue(name: String, attrs: List<Attribute>) = attrs.firstOrNull { it.name == name } ?: emptyAttribute
```

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt [hunk 3]
```diff
@@ -7,26 +7,45 @@ import ac.uk.ebi.biostd.submission.File
...
+        elements.add(element)
+    }
+}
+
+data class TableHeader(val name: String, val termNames: List<String> = listOf()) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || javaClass != other.javaClass) return false
+        val that = other as TableHeader?
+        return name == that!!.name
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: can be replaced with `return name == (other as TableHeader).name`

Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: sure

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt (hunk 2 above)
Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: sure. I've added Attribute.EMPTY.. please have a look

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt [hunk 4]
```diff
@@ -7,33 +7,55 @@ import ac.uk.ebi.biostd.submission.File
...
+
+    override fun hashCode(): Int {
+        return name.hashCode()
     }
 }
 
+data class TableRow(val id: String, val idPropertyName: String, val values: List<Attribute>) {
+    fun valueList(): List<String> = listOf(id) +
+            values.flatMap { attr ->
+                if (attr == Attribute.EMPTY) listOf("") else listOf(attr.value) + attr.terms.map { it.second }
```
[the remark this answers is not in the document]
Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]: sure. missed it.. thnks.

**Pull request merged**